### PR TITLE
Assessment hierarchy

### PR DESCRIPTION
**Pull request by nategbrown9**
Add support for loading an assessment containing other assessments.

### CODE REVIEW

File: assessmentModel/src/commonMain/kotlin/serialization/Serialization.kt [hunk 1]
```diff
@@ -30,6 +32,38 @@ object Serialization {
...
+    override fun getResourceInfo(assessmentIdentifier: String): ResourceInfo {
+        return defaultResourceInfo
+    }
+
+    override fun getJsonDecoder(assessmentIdentifier: String?): Json {
+        return Serialization.JsonCoder.default
+    }
+
+}
+
```
Comment (syoung-smallwisdom) [inline assessmentModel/src/commonMain/kotlin/serialization/Serialization.kt:66]: I would suggest the following:

First, refactor `Assessment` to pull out the bits which come from the server.

```
interface AssessmentInfo {

    /**
     * The identifier for the assessment.
     */
    val identifier: String

    /**
     * The [versionString] may be a semantic version, timestamp, or sequential revision integer.
     */
    val versionString: String?

    /**
     * The [schemaIdentifier] is used to allow a group of [Assessment] models to all map to the same
     * table in a data base.
     */
    val schemaIdentifier: String?
}
```

Which makes the Assessment look like this:
```
interface Assessment : BranchNode, ContentNode, AssessmentInfo
``` 

Then change the `ModuleInfoProvider` to be:
```
/**
 * The [ModuleInfoProvider] provides methods to lookup the necessary [ResourceInfo] and [Json] to
 * load an [Assessment] using the provide the platform specific [FileLoader].
 */
interface ModuleInfoProvider {

    val fileLoader: FileLoader

    fun getRegisteredResourceInfo(assessmentInfo: AssessmentInfo): ResourceInfo? = null

    fun getRegisteredJsonDecoder(assessmentInfo: AssessmentInfo): Json? = null

    /**
     * Allows the provider to return an [Assessment] (for the given [assessmentInfo]) that is not
     * encoded as a JSON serialized object.
     */
    fun getRegisteredAssessment(assessmentInfo: AssessmentInfo): Assessment? = null
}
```

And finally:
```
/**
 * A [TransformableAssessment] is a placeholder that can be used to contain just enough information about an
 * [Assessment] to allow referencing the resource information needed to load the actual [Assessment] on demand. This
 * allows using a "placeholder" that can be vended from a different source from the actual assessment. For example,
 * an active task may be defined using a hardcoded JSON file and included in a module but requested via a
 * [TransformableAssessment] that is vended from a server.
 */
interface TransformableAssessment : Assessment, TransformableNode {
    override fun unpack(moduleInfoProvider: ModuleInfoProvider, resourceInfo: ResourceInfo, jsonCoder: Json): Assessment {
        val assessment = moduleInfoProvider.getRegisteredAssessment(this)
        return if (assessment != null) assessment else {
            val curResourceInfo = moduleInfoProvider.getRegisteredResourceInfo(this) ?: resourceInfo
            val serializer = PolymorphicSerializer(Assessment::class)
            val jsonString = moduleInfoProvider.fileLoader.loadFile(this, resourceInfo)
            val curJsonCoder = moduleInfoProvider.getRegisteredJsonDecoder(this) ?: jsonCoder
            val unpackedNode = curJsonCoder.decodeFromString(serializer, jsonString)
            unpackedNode.unpack(moduleInfoProvider.fileLoader, curResourceInfo, curJsonCoder)
        }
    }

    override fun getNavigator(): Navigator = throw IllegalStateException("Attempted to get a navigator without first unpacking the Assessment.")
}
```

File: assessmentModel/src/commonMain/kotlin/serialization/FileAssessmentProvider.kt [hunk 2]
```diff
@@ -39,14 +42,17 @@ interface TransformableNode : ContentNode, AssetInfo {
...
+
+    override fun unpack(moduleInfoProvider: ModuleInfoProvider, resourceInfo: ResourceInfo, jsonCoder: Json): Assessment {
+        val curResourceInfo = moduleInfoProvider.getResourceInfo(this.identifier)
         val serializer = PolymorphicSerializer(Assessment::class)
-        val jsonString = fileLoader.loadFile(this, resourceInfo)
+        val jsonString = moduleInfoProvider.fileLoader.loadFile(this, resourceInfo)
+        val curJsonCoder = moduleInfoProvider.getJsonDecoder(this.identifier)
         val unpackedNode = jsonCoder.decodeFromString(serializer, jsonString)
-        return unpackedNode.unpack(fileLoader, resourceInfo, jsonCoder)
+        return unpackedNode.unpack(moduleInfoProvider, curResourceInfo, curJsonCoder)
```
Comment (syoung-smallwisdom) [inline assessmentModel/src/commonMain/kotlin/serialization/FileAssessmentProvider.kt:52]: val unpackedNode = **_curJsonCoder._**decodeFromString(serializer, jsonString)

**Review by syoung-smallwisdom — CHANGES_REQUESTED** [review]
A few tweeks to suggest.

**Review by syoung-smallwisdom — APPROVED** [review]
Some nits and a PR with suggested changes.

File: assessmentModel/src/commonMain/kotlin/navigation/NodeState.kt [hunk 3]
```diff
@@ -400,7 +430,9 @@ open class BranchNodeStateImpl(override val node: BranchNode, final override val
      * class looks for conformance to the [BranchNode] interface.
      */
     open fun getBranchNodeState(navigationPoint: NavigationPoint): BranchNodeState? {
-        return (navigationPoint.node as? BranchNode)?.let { BranchNodeStateImpl(it, this) }
+        return (navigationPoint.node as? BranchNode)?.let {
+            customBranchNodeStateProvider?.customNodeStateFor(it, this) ?: BranchNodeStateImpl(it, this)
```
[the remark this answers is not in the document]
Comment (nategbrown9) [reply assessmentModel/src/commonMain/kotlin/navigation/NodeState.kt:433]: I think for developers that don't need a custom `BranchNodeState` implementation, it is easier to keep them separate. I changed the method name on the `BranchNodeState` interface to `customLeafNodeStateFor(node: Node, parent: BranchNodeState)' to be more explicit.

File: assessmentModel/src/commonMain/kotlin/ModuleInfo.kt [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+     * Returns the [Assessment] that is unpacked by the best match of a [TransformableAssessment]
+     * that is contained within this module.
+     */
+    override fun getAssessment(
+        assessmentPlaceholder: AssessmentPlaceholder,
+        registryProvider: AssessmentRegistryProvider
+    ): Assessment {
+        val sublist = assessments.filter { it.identifier == assessmentPlaceholder.assessmentInfo.identifier }
+        // TODO: syoung 01/27/2021 Look at version and schema identifier
+        return sublist.first().unpack(assessmentPlaceholder,this, registryProvider)
```
Comment (syoung-smallwisdom) [inline assessmentModel/src/commonMain/kotlin/ModuleInfo.kt:90]:
```suggestion
        return sublist.first().unpack(assessmentPlaceholder, this, registryProvider)
```
Oops.

File: assessmentModel/src/commonMain/kotlin/navigation/NodeState.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (syoung-smallwisdom) [reply assessmentModel/src/commonMain/kotlin/navigation/NodeState.kt:433]: I guess I'm confused as to why an assessment developer may want to use a custom BranchNodeState but *not* want to use that same pattern for the leaf node states. Especially given that you are separating the concerns of "handle finished" and "present UI". Why wouldn't the custom leaf state *also* be vended by the same provider (if it isn't handled by the branch node state using override)?

I'm wondering if maybe the same kind of pair-programming day with Ella would be a good idea?

File: assessmentModel/src/commonMain/kotlin/serialization/Nodes.kt [hunk 5]
```diff
@@ -124,6 +121,29 @@ abstract class IconNodeObject : NodeObject() {
     }
 }
 
+/**
+ * A concrete implementation of an [AssessmentPlaceholder].
+ * TODO: syoung 01/27/2021 Add unit test of the serialization.
```
Comment (syoung-smallwisdom) [inline assessmentModel/src/commonMain/kotlin/serialization/Nodes.kt:126]:
```suggestion
```

This is done. Forgot to remove the TODO comment.

**Review by syoung-smallwisdom — APPROVED** [review]
A little cleanup and a question.

Otherwise, looks good.

File: assessmentModel/src/commonMain/kotlin/navigation/NodeState.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (nategbrown9) [reply assessmentModel/src/commonMain/kotlin/navigation/NodeState.kt:433]: I took a look at the MobileToolBox code, and they are reverse of what I would have argued would be the common case. They have a custom `BranchNodeState` implementation, but no custom leaf NodeStates.

**Pull request merged**